File: server/src/NomNomzBot.Api/Controllers/V1/ChannelBotController.cs
// -----------------------------------------------------------------------------
//  Copyright (c) NoMercy Labs.
//
//  This file is part of NomNomzBot, free software licensed under the GNU Affero
//  General Public License v3.0 or later. You may redistribute and/or modify it
//  under those terms. Distributed WITHOUT ANY WARRANTY. See LICENSE for details.
//
//  SPDX-License-Identifier: AGPL-3.0-or-later
// -----------------------------------------------------------------------------

using Asp.Versioning;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.AspNetCore.RateLimiting;
using NomNomzBot.Api.Authorization;
using NomNomzBot.Api.Extensions;
using NomNomzBot.Api.Models;
using NomNomzBot.Application.Common.Models;
using NomNomzBot.Application.Identity.Dtos;
using NomNomzBot.Application.Identity.Services;
using NomNomzBot.Application.Integrations.Dtos;
using NomNomzBot.Domain.Identity.Enums;

namespace NomNomzBot.Api.Controllers.V1;

/// <summary>
/// Manages white-label (per-channel) bot accounts.
/// A white-label bot is a separate Twitch account the channel owner authenticates so
/// that bot messages appear from their own bot identity rather than the platform bot (NomNomzBot).
///
/// Token is stored as Service(Name="twitch_bot", BroadcasterId=channelId).
/// The platform bot (BroadcasterId=null) is managed in AdminController.
/// </summary>
[ApiVersion("1.0")]
[Route("api/v{version:apiVersion}/channels")]
[Tags("Channel Bot")]
public class ChannelBotController : BaseController
{
    // The streamer's own Twitch connection (self-host: the bot chats as the streamer's own account until a
    // dedicated bot registers) and the dedicated white-label bot connection (twitch_bot). Mirrors
    // AuthService's TwitchProvider / PlatformBotProvider so the scopes page reads exactly what the bot flow writes.
    private const string TwitchProvider = AuthEnums.IntegrationProvider.Twitch;
    private const string TwitchBotProvider = AuthEnums.IntegrationProvider.Twitch + "_bot";

    private readonly IAuthService _authService;
    private readonly IIntegrationTokenVault _vault;
    private readonly IConfiguration _config;
    private readonly ITwitchOAuthStateService _oauthState;

    public ChannelBotController(
        IAuthService authService,
        IIntegrationTokenVault vault,
        IConfiguration config,
        ITwitchOAuthStateService oauthState
    )
    {
        _authService = authService;
        _vault = vault;
        _config = config;
        _oauthState = oauthState;
    }

    private string GetPublicBaseUrl() => Request.ResolvePublicOrigin(_config);

    public record ScopeDto(
        string Scope,
        string Name,
        string Description,
        string Category,
        bool Granted,
        bool Required
    );

    public record ScopesResponseDto(List<ScopeDto> Permissions, int GrantedCount, int TotalCount);

    // The scopes the white-label bot OAuth flow actually grants (mirrors AuthService.BotScopes). This is the
    // BOT account's chat-permission page, not the streamer's — so it lists only the bot's chat scopes.
    //  - user:read:chat / user:write:chat are the live Helix chat path (read via EventSub, send via
    //    POST /helix/chat/messages) and are REQUIRED: their absence is a genuine "action required".
    //  - chat:read / chat:edit are the legacy IRC scopes. The bot OAuth still requests them, but the Helix
    //    chat path never exercises them (IRC is fully retired, scaling-qos.md §6). They are listed for
    //    transparency but are NOT required — so a self-host bot chatting as the streamer's own account
    //    (which holds the Helix scopes but not the IRC ones) still reads as complete.
    //  - user:read:whispers feeds the bot's whisper inbox (the platform-plane user.whisper.message topic).
    //    Not required: its absence 403s only that one topic, and no whisper-consuming surface exists yet —
    //    it flips to required when the bot-inbox feature ships.
    private static readonly (
        string Scope,
        string Name,
        string Description,
        string Category,
        bool Required
    )[] KnownScopes =
    [
        (
            "user:read:chat",
            "Read Chat",
            "Read chat messages so the bot can see and respond to commands",
            "Chat",
            true
        ),
        ("user:write:chat", "Send Chat Messages", "Send chat messages as the bot", "Chat", true),
        (
            "chat:read",
            "Read Chat (legacy IRC)",
            "Legacy IRC chat read — superseded by the Helix chat path and no longer used",
            "Chat",
            false
        ),
        (
            "chat:edit",
            "Send Chat (legacy IRC)",
            "Legacy IRC chat send — superseded by the Helix chat path and no longer used",
            "Chat",
            false
        ),
        (
            "user:read:whispers",
            "Read Whispers",
            "Receive whispers sent to the bot account (the bot's whisper inbox)",
            "Whispers",
            false
        ),
    ];

    /// <summary>
    /// Returns the white-label bot account's chat-permission status for this channel — the signal behind the
    /// dashboard's bot "action required" prompt. Reads the dedicated bot connection (<c>twitch_bot</c>, written
    /// by the bot OAuth flow), falling back to the streamer's own <c>twitch</c> connection for the self-host case
    /// where the bot chats as the streamer's own account until a custom bot is registered. Only the required
    /// scopes gate completion, so once the bot grant lands the prompt clears (100%).
    /// </summary>
    [HttpGet("{channelId}/scopes")]
    [Authorize]
    [RequireAction("channelbot:read")]
    [ProducesResponseType<StatusResponseDto<ScopesResponseDto>>(StatusCodes.Status200OK)]
    public async Task<IActionResult> GetScopes(string channelId, CancellationToken ct)
    {
        if (!Guid.TryParse(channelId, out Guid tenantId))
            return BadRequestResponse("Invalid channel id.");

        Result<IReadOnlyList<IntegrationConnectionDto>> connections =
            await _vault.ListConnectionsAsync(tenantId, ct);
        IReadOnlyList<IntegrationConnectionDto> list = connections.IsSuccess
            ? connections.Value
            : [];

        // Prefer the dedicated white-label bot connection; fall back to the streamer's own connection for the
        // self-host case where the bot IS the streamer's own account (bot-identity self-host fallback).
        IntegrationConnectionDto? botConnection =
            list.FirstOrDefault(c => c.Provider == TwitchBotProvider)
            ?? list.FirstOrDefault(c => c.Provider == TwitchProvider);

        HashSet<string> grantedScopes =
            botConnection?.Scopes.ToHashSet(StringComparer.OrdinalIgnoreCase)
            ?? new HashSet<string>(StringComparer.OrdinalIgnoreCase);

        List<ScopeDto> permissions = KnownScopes
            .Select(s => new ScopeDto(
                s.Scope,
                s.Name,
                s.Description,
                s.Category,
                grantedScopes.Contains(s.Scope),
                s.Required
            ))
            .ToList();

        // Only the required scopes decide the "action required" prompt. The legacy IRC scopes are informational
        // (Required=false) and the Helix chat path never exercises them, so their absence must never keep the
        // prompt open — a completed bot grant (or a self-host streamer token carrying the Helix chat scopes)
        // reads 100%.
        int totalCount = permissions.Count(p => p.Required);
        int grantedCount = permissions.Count(p => p.Required && p.Granted);

        return Ok(
            new StatusResponseDto<ScopesResponseDto>
            {
                Data = new ScopesResponseDto(permissions, grantedCount, totalCount),
            }
        );
    }

    /// <summary>
    /// Start Twitch OAuth for this channel's white-label bot — an additional bot-identity registration on top
    /// of the owner's account (not a login), handled like an integration connect. Gated owner-level
    /// (identity-auth §5, <c>management/Broadcaster · channelbot:connect</c>). Returns the Twitch authorize URL
    /// (force_verify) for the client to open; the callback completes via the unified, nonce-validated
    /// <c>/auth/twitch/callback</c>.
    /// </summary>
    [HttpGet("{channelId}/bot/connect")]
    [Authorize]
    [RequireAction("channelbot:connect")]
    [EnableRateLimiting("auth")]
    public async Task<IActionResult> StartChannelBotOAuth(string channelId, CancellationToken ct)
    {
        if (!Guid.TryParse(channelId, out Guid tenantId))
            return BadRequestResponse("Invalid channel id.");

        // Issue a single-use, server-side CSRF state nonce holding the flow + tenant id, so the channel-bot
        // association cannot be forged by tampering with the state (§5). Only the opaque nonce leaves us.
        string state = await _oauthState.IssueAsync(
            new TwitchOAuthFlowState("channel_bot", ChannelId: channelId),
            ct
        );

        Result<string> authUrl = await _authService.GetTwitchChannelBotOAuthUrl(
            tenantId,
            state,
            baseUrl: GetPublicBaseUrl(),
            cancellationToken: ct
        );
        if (authUrl.IsFailure)
            return ResultResponse(authUrl);
        return Ok(
            new StatusResponseDto<OAuthStartDto> { Data = new OAuthStartDto(authUrl.Value, state) }
        );
    }

    /// <summary>Get white-label bot status for a specific channel.</summary>
    [HttpGet("{channelId}/bot/status")]
    [Authorize]
    [RequireAction("channelbot:read")]
    public async Task<IActionResult> GetChannelBotStatus(string channelId, CancellationToken ct)
    {
        if (!Guid.TryParse(channelId, out Guid tenantId))
            return BadRequestResponse("Invalid channel id.");

        Result<BotStatusDto> result = await _authService.GetChannelBotStatusAsync(tenantId, ct);
        return ResultResponse(result);
    }

    /// <summary>Disconnect the white-label bot for a specific channel.</summary>
    [HttpDelete("{channelId}/bot")]
    [Authorize]
    [RequireAction("channelbot:disconnect")]
    public async Task<IActionResult> DisconnectChannelBot(string channelId, CancellationToken ct)
    {
        if (!Guid.TryParse(channelId, out Guid tenantId))
            return BadRequestResponse("Invalid channel id.");

        Result result = await _authService.DisconnectChannelBotAsync(tenantId, ct);
        return ResultResponse(result);
    }
}

File: app/composeApp/src/commonMain/kotlin/bot/nomnomz/dashboard/feature/settings/state/ChannelBotController.kt
// -----------------------------------------------------------------------------
//  Copyright (c) NoMercy Labs.
//
//  This file is part of NomNomzBot, free software licensed under the GNU Affero
//  General Public License v3.0 or later. You may redistribute and/or modify it
//  under those terms. Distributed WITHOUT ANY WARRANTY. See LICENSE for details.
//
//  SPDX-License-Identifier: AGPL-3.0-or-later
// -----------------------------------------------------------------------------

package bot.nomnomz.dashboard.feature.settings.state

import bot.nomnomz.dashboard.core.network.ApiResult
import bot.nomnomz.dashboard.core.network.ChannelBotStatusDetail
import bot.nomnomz.dashboard.core.network.ChannelScope
import bot.nomnomz.dashboard.core.network.ChannelScopesResponse
import bot.nomnomz.dashboard.core.network.ChannelSummary
import bot.nomnomz.dashboard.core.network.ChannelsApi
import bot.nomnomz.dashboard.core.network.OAuthStart
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

// State-holder for the channel's white-label bot card in Settings. The channel bot is an OPTIONAL
// dedicated Twitch account whose messages appear from a channel-specific identity instead of the
// shared platform bot. Connecting it opens a Twitch OAuth flow; disconnecting revokes the stored
// token without affecting the channel record or the fallback platform bot.
class ChannelBotController(private val channelsApi: ChannelsApi) {
    private val _state: MutableStateFlow<ChannelBotState> = MutableStateFlow(ChannelBotState.Loading)

    /** The card's render state. */
    val state: StateFlow<ChannelBotState> = _state.asStateFlow()

    private var channelId: String? = null

    /** Resolve the active channel, then load the bot status and the broadcaster-token scope list. */
    suspend fun load() {
        _state.value = ChannelBotState.Loading

        val channel: ChannelSummary =
            when (val result: ApiResult<ChannelSummary> = channelsApi.primaryChannel()) {
                is ApiResult.Failure -> {
                    _state.value = ChannelBotState.Error(result.error.message)
                    return
                }
                is ApiResult.Ok -> result.value
            }

        channelId = channel.id

        val botStatus: ChannelBotStatusDetail =
            when (val result: ApiResult<ChannelBotStatusDetail> = channelsApi.channelBotStatus(channel.id)) {
                is ApiResult.Failure -> {
                    _state.value = ChannelBotState.Error(result.error.message)
                    return
                }
                is ApiResult.Ok -> result.value
            }

        val scopes: List<ChannelScope> =
            when (val result: ApiResult<ChannelScopesResponse> = channelsApi.channelScopes(channel.id)) {
                is ApiResult.Failure -> emptyList()
                is ApiResult.Ok -> result.value.permissions
            }

        _state.value =
            ChannelBotState.Ready(
                connected = botStatus.connected,
                login = botStatus.displayName ?: botStatus.login,
                scopes = scopes,
            )
    }

    /**
     * Start the Twitch OAuth flow for the channel's white-label bot. Returns the authorization URL
     * the caller should open in the system browser; null when the channel is not resolved or the
     * backend call fails. The bot status refreshes automatically when Twitch redirects back.
     */
    suspend fun startConnect(): String? {
        val target: String = channelId ?: return null
        return when (val result: ApiResult<OAuthStart> = channelsApi.startChannelBotConnect(target)) {
            is ApiResult.Failure -> {
                applyActionError(result.error.message)
                null
            }
            is ApiResult.Ok -> result.value.authorizeUrl
        }
    }

    /** Revoke and remove the channel's white-label bot account, then reload the status. */
    suspend fun disconnect() {
        val target: String = channelId ?: return
        val current: ChannelBotState = _state.value
        if (current !is ChannelBotState.Ready) return
        _state.value = current.copy(busy = true, actionError = null)
        when (val result: ApiResult<Unit> = channelsApi.disconnectChannelBot(target)) {
            is ApiResult.Ok -> load()
            is ApiResult.Failure -> applyActionError(result.error.message)
        }
    }

    private fun applyActionError(message: String) {
        val current: ChannelBotState = _state.value
        _state.value =
            if (current is ChannelBotState.Ready) current.copy(busy = false, actionError = message)
            else current
    }
}

/** The channel white-label bot card's render state. */
sealed interface ChannelBotState {
    data object Loading : ChannelBotState

    data class Ready(
        val connected: Boolean,
        val login: String?,
        val scopes: List<ChannelScope> = emptyList(),
        val busy: Boolean = false,
        val actionError: String? = null,
    ) : ChannelBotState

    data class Error(val detail: String) : ChannelBotState
}
